nexus/db-model/src/network_interface.rs

```diff
@@ -14,36 +17,184 @@ use nexus_types::identity::Resource;
+    #[diesel(postgres_type(name = "network_interface_kind"))]
+    pub struct NetworkInterfaceKindEnum;
+
+    #[derive(Clone, Copy, Debug, AsExpression, FromSqlRow, PartialEq)]
+    #[diesel(sql_type = NetworkInterfaceKindEnum)]
+    pub enum NetworkInterfaceKind;
+
+    Instance => b"instance"
+    Service => b"service"
+}
+
+/// Generic Network Interface DB model.
+#[derive(Selectable, Queryable, Clone, Debug, Resource)]
 #[diesel(table_name = network_interface)]
 pub struct NetworkInterface {
     #[diesel(embed)]
     pub identity: NetworkInterfaceIdentity,
 
-    pub instance_id: Uuid,
+    pub kind: NetworkInterfaceKind,
+    pub parent_id: Uuid,
+
     pub vpc_id: Uuid,
     pub subnet_id: Uuid,
+
     pub mac: MacAddr,
     // TODO-correctness: We need to split this into an optional V4 and optional V6 address, at
     // least one of which will always be specified.
     //
     // If user requests an address of either kind, give exactly that and not the other.
     // If neither is specified, auto-assign one of each?
     pub ip: ipnetwork::IpNetwork,
+
     pub slot: i16,
     #[diesel(column_name = is_primary)]
     pub primary: bool,
 }
 
-impl From<NetworkInterface> for external::NetworkInterface {
-    fn from(iface: NetworkInterface) -> Self {
-        Self {
-            identity: iface.identity(),
-            instance_id: iface.instance_id,
+/// Instance Network Interface DB model.
+///
+/// The underlying "table" (`instance_network_interface`) is actually a view
+/// over the `network_interface` table, that contains only rows with
+/// `kind = 'instance'`.
+#[derive(Selectable, Queryable, Clone, Debug, Resource)]
+#[diesel(table_name = instance_network_interface)]
+pub struct InstanceNetworkInterface {
+    #[diesel(embed)]
+    pub identity: InstanceNetworkInterfaceIdentity,
+
+    pub instance_id: Uuid,
+    pub vpc_id: Uuid,
+    pub subnet_id: Uuid,
+
+    pub mac: MacAddr,
+    pub ip: ipnetwork::IpNetwork,
+
+    pub slot: i16,
+    #[diesel(column_name = is_primary)]
+    pub primary: bool,
+}
+
+/// Service Network Interface DB model.
+///
+/// The underlying "table" (`service_network_interface`) is actually a view
+/// over the `network_interface` table, that contains only rows with
+/// `kind = 'service'`.
+#[derive(Selectable, Queryable, Clone, Debug, Resource)]
+#[diesel(table_name = service_network_interface)]
+pub struct ServiceNetworkInterface {
+    #[diesel(embed)]
+    pub identity: ServiceNetworkInterfaceIdentity,
+
+    pub service_id: Uuid,
+    pub vpc_id: Uuid,
+    pub subnet_id: Uuid,
+
+    pub mac: MacAddr,
+    pub ip: ipnetwork::IpNetwork,
+
+    pub slot: i16,
+    #[diesel(column_name = is_primary)]
+    pub primary: bool,
+}
+
+impl NetworkInterface {
+    /// Treat this `NetworkInterface` as an `InstanceNetworkInterface`.
+    ///
+    /// # Panics
+    /// Panics if this isn't an 'instance' kind network interface.
+    pub fn as_instance(self) -> InstanceNetworkInterface {
```

Comment:
Is there a reason not to do a `TryFrom` instead?

Reply:
Hmm, mostly just cause the only places I used it expect the conversion to be infallible and so using `TryFrom`/`TryInto` would just mean shuffling the panic/unwrap.